src/app/core/services/wizard/provider/openstack.spec.ts

```diff
@@ -0,0 +1,259 @@
+        });
+
+      const req = httpMock.expectOne(service.availabilityZonesUrl);
+      expect(req.request.method).toBe('GET');
+      expect_projectID_auth_header_to_be_defined(req);
+      req.flush(dummyAZ);
+    });
+
+    test('with ApplicationID auth', () => {
+      service.datacenter(FakeValue.Datacenter);
+      service.applicationCredentialID(FakeValue.ApplicationCredentialID);
+      service.applicationCredentialPassword(FakeValue.ApplicationCredentialSecret);
+      service
+        .availabilityZones(() => {})
+        .subscribe(az => {
+          expect(az.length).toBe(1);
+          expect(az).toEqual(dummyAZ);
+        });
+
+      const req = httpMock.expectOne(service.availabilityZonesUrl);
+      expect(req.request.method).toBe('GET');
+      expect_applicationId_header_to_be_defined(req);
+      req.flush(dummyAZ);
+    });
+  });
+
+  describe('#securityGroups', () => {
+    const dummySecGroup: OpenstackSecurityGroup[] = [
+      {
+        name: 'sec-group--name',
+        id: 'sec-group-id',
+      },
+    ];
+
+    test('with project auth', () => {
+      service.project(FakeValue.Project);
+
+      service
+        .securityGroups(() => {})
+        .subscribe(secGroup => {
+          expect(secGroup.length).toBe(1);
+          expect(secGroup).toEqual(dummySecGroup);
+        });
+
+      const req = httpMock.expectOne(service.securityGroupsUrl);
+      expect(req.request.method).toBe('GET');
+      expect_project_auth_header_to_be_defined(req);
+      req.flush(dummySecGroup);
+    });
+    test('with projectID auth', () => {
+      service.projectID(FakeValue.ProjectID);
+
+      service
+        .securityGroups(() => {})
+        .subscribe(secGroup => {
+          expect(secGroup.length).toBe(1);
+          expect(secGroup).toEqual(dummySecGroup);
+        });
+
+      const req = httpMock.expectOne(service.securityGroupsUrl);
+      expect(req.request.method).toBe('GET');
+      expect_projectID_auth_header_to_be_defined(req);
+      req.flush(dummySecGroup);
+    });
+
+    test('with ApplicationID auth', () => {
+      service.datacenter(FakeValue.Datacenter);
+      service.applicationCredentialID(FakeValue.ApplicationCredentialID);
+      service.applicationCredentialPassword(FakeValue.ApplicationCredentialSecret);
+      service
+        .securityGroups(() => {})
+        .subscribe(secGroup => {
+          expect(secGroup.length).toBe(1);
+          expect(secGroup).toEqual(dummySecGroup);
+        });
+
+      const req = httpMock.expectOne(service.securityGroupsUrl);
+      expect(req.request.method).toBe('GET');
+      expect_applicationId_header_to_be_defined(req);
+      req.flush(dummySecGroup);
+    });
+  });
+});
+
+enum FakeValue {
+  Username = 'the-user',
+  Password = 'fake-pass',
+  ApplicationCredentialID = 'the-application-id',
+  ApplicationCredentialSecret = 'the-ApplicationCredentialSecret',
+  Domain = 'the-domain',
+  Datacenter = 'the-datacenter',
+  Project = 'the-roject',
```

Comment:
It doesn't change anything but "p" is missing :)

Reply:
nice catch :)